### PR TITLE
Enable encrypted HCL files via Keybase

### PR DESCRIPTION
**Pull request by mmartell37**
- Enable using Keybase with one or more PGP keys to decrypt and encrypt arbitrary files.
- Enable editing arbitrary files.
- Enable processing both HCL (clear text) and PGP (encrypted) files automatically.
- Enable remote `vault-list-secrets`.
- Add `environment` prefix when writing secrets.
- Enable remote secrets deletion.

### CODE REVIEW

**Comment by josegonzalez** [comment]
Seems like a lot of code. For reference, we use an external tool to unpack/repack secrets via AWS KMS before doing work - and also checking the diff before merging a pull request - but I defer to @jippi if he wants this in core/any general thoughts :)

File: README.md [hunk 1]
```diff
@@ -31,6 +35,23 @@ The following environment variables are required for setting configuration and k
...
+vault_profile_name_2:
+  server: http://<ip>:8200
+  token: <your token>
+...
+encrypted-file-config:
+  keybase-team-name: < your keybase team name >
+  editor: < your favorite editor > # defaults to $EDITOR, this, then pico
+...
+```
+Note that if `--no-keybase-team` global flag is set, `keybase-team-name` will be ignored and only the local Keybase PGP profile will be used.
```
Comment (jippi) [inline README.md]: if `keybase-team-name` is not set in yaml, will that be the same as `--no-keybase-team`?

File: README.md [hunk 2]
```diff
@@ -45,18 +66,28 @@ hashi-helper [--global-flags] command [--command-flags]
 - `--config-file` / `CONFIG_FILE`: A single `hcl` configuration file to parse instead of a directory (optional; default: `<empty>`)
 - `--environment` / `ENVIRONMENT`: The environment to process for (optional; default: `all`)
 - `--application` / `APPLICATION`: The application to process for (optional; default: `all`)
+- `--warn-unencrypted` / `WARN_UNENCRYPTED`: Issue a warning if unencrypted HCL files are discovered.
+- `--no-keybase-team` / `NO_KEYBASE_TEAM`: Ignore the Keybase team name in `.hashi-helper-config.yml.pgp` (if set).
```
Comment (jippi) [inline README.md]: maybe make this into `KEYBASE_TEAM`, and if the ENV, CLI flag or the YAML value is empty, then behave like `--no-keybase-team` ?

File: command/edit.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+	cli "gopkg.in/urfave/cli.v1"
+	log "github.com/Sirupsen/logrus"
+	yaml "gopkg.in/yaml.v2"
+	vault "github.com/seatgeek/hashi-helper/command/vault"
+)
+
+type preferences map[string]preference
+type preference struct {
+	KeybaseTeam  string `yaml:"keybase-team-name"`
```
Comment (jippi) [inline command/edit.go]: can you `go fmt` all the files please?

File: command/edit.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+	if !ok {
+		return fmt.Errorf("No encrypted-file-config block found in config file.\n")
+	}
+	// editor
+	editor := os.Getenv("EDITOR")
+	if editor == "" {
+		editor = preference.Editor
+	}
+	if editor == "" {
+		editor = "pico"
```
Comment (jippi) [inline command/edit.go]: i would error if there is no `EDITOR` rather than default to `pico` (this won't work on Windows for example)

File: command/edit.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+	}
+	// keybase team
+	var keybaseTeam string = preference.KeybaseTeam
+	if keybaseTeam == "" && NoKeybaseTeam == false {
+		log.Warnf("No Keybase Team found in config (and --no-keybase-team is unset). Not using Keybase Team.\n")
```
Comment (jippi) [inline command/edit.go]: this might be desired, not sure if it should be a warning? 

If its not defined in any configuration source, the desired behaviour might be to use local keybase instead

File: command/edit.go [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+		encryptCmd = exec.Command("bash", "-c", cmd)
+	  log.Infof("Using Keybase team \"%s\" with members \"%s\".\n", keybaseTeam, memberList)
+	}
+	encryptErr := encryptCmd.Run()
+	if encryptErr != nil {
+		return encryptErr
+	}
+
+	// defered action to remove temporary file post-edit and post-re-encryption
+	defer os.Remove(tempFile.Name())
```
Comment (jippi) [inline command/edit.go]: this should be moved up to line 69

File: command/vault/helper/secret_writer.go [hunk 7]
```diff
@@ -17,14 +17,14 @@ type SecretWriter struct {
...
-func (w SecretWriter) WriteSecret(secret *config.Secret) error {
+func (w SecretWriter) WriteSecret(secret *config.Secret, env string) error {
 	var path string
 
 	// @TODO Make a dedicated type for writing non-secrets !
 	if strings.HasPrefix(secret.Path, "/") {
 		path = strings.TrimLeft(secret.Path, "/")
 	} else if secret.Application != nil {
-		path = fmt.Sprintf("secret/%s/%s", secret.Application.Name, secret.Path)
+		path = fmt.Sprintf("secret/%s/%s/%s", env, secret.Application.Name, secret.Path)
```
Comment (jippi) [inline command/vault/helper/secret_writer.go]: this is a breaking, unrelated change to your PR title, please open a new PR for discussion on this

File: command/vault/helper/secret_deleter.go [hunk 8]
```diff
@@ -0,0 +1,45 @@
+package helper
```
Comment (jippi) [inline command/vault/helper/secret_deleter.go]: this is a breaking, unrelated change to your PR title, please open a new PR for discussion on this

File: command/vault/secrets_delete.go [hunk 9]
```diff
@@ -0,0 +1,75 @@
+package vault
```
Comment (jippi) [inline command/vault/secrets_delete.go]: this is a breaking, unrelated change to your PR title, please open a new PR for discussion on this

File: command/vault/secrets_pull.go [hunk 10]
```diff
@@ -95,14 +95,14 @@ func SecretsPull(c *cli.Context) error {
 			if err != nil {
```
Comment (jippi) [inline command/vault/secrets_pull.go]: this is a breaking, unrelated change to your PR title, please open a new PR for discussion on this

File: command/vault/secrets_list.go [hunk 11]
```diff
@@ -10,7 +10,7 @@ import (
```
Comment (jippi) [inline command/vault/secrets_list.go]: this is a breaking, unrelated change to your PR title, please open a new PR for discussion on this

File: command/vault/secrets_import.go [hunk 12]
```diff
@@ -79,14 +79,14 @@ func SecretsImport(c *cli.Context) error {
 			if err != nil {
```
Comment (jippi) [inline command/vault/secrets_import.go]: this is a breaking, unrelated change to your PR title, please open a new PR for discussion on this

File: command/vault/secrets_push.go [hunk 13]
```diff
@@ -32,7 +32,7 @@ func SecretsPushWithConfig(c *cli.Context, config *config.Config) error {
```
Comment (jippi) [inline command/vault/secrets_push.go]: this is a breaking, unrelated change to your PR title, please open a new PR for discussion on this

File: config/type_config.go [hunk 14]
```diff
@@ -1,9 +1,11 @@
 package config
```
Comment (jippi) [inline config/type_config.go]: `go fmt' this file

**Comment by mmartell37** [comment]
@josegonzalez, @jippi thanks for the feedback and the thorough review! This is actually my first public PR, and my first foray into go. It is a lot of code, and I'll do my best to break it up. Thanks for your patience :)

File: command/edit.go (hunk 4 above)
Comment (mmartell37) [reply command/edit.go]: 👍 Sounds good.

File: command/edit.go (hunk 5 above)
Comment (mmartell37) [reply command/edit.go]: Good thought. I was only thinking of my particular use case (signing with multiple PGP keys [a team] is required). But in general I suppose that is not the case. Combined with your comment above w.r.t. `KEYBASE_TEAM`, I agree that will make for a more sane set of default behaviors.

Comment (mmartell37) [reply command/edit.go]: Good thought. I was only thinking of my particular use case (signing with multiple PGP keys [a team] is required). But in general I suppose that is not the case. Combined with your comment above w.r.t. KEYBASE_TEAM, I agree that will make for a more sane set of default behaviors.

File: command/edit.go (hunk 3 above)
Comment (mmartell37) [reply command/edit.go]: Yes, apologies. First experience with go.

**Comment by jippi** [comment]
@mmartell37 did you stall on this? :)

**Comment by mmartell37** [comment]
@jippi I did, apologies. I switched gears to a moderately urgent project and haven't had time to get back to this yet.

**Comment by jippi** [comment]
alright, no problem :)

I'm going to close this PR, as it will be revived in smaller chunks if you have time at a later point :)
